### PR TITLE
[Clang] [CodeGen] Fix codegen bug in constant initialisation in C23 mode

### LINKED ISSUE #84784: C23 mode: cannot compile this static initializer yet
In C23, clang as of trunk gives me that error on the following code

```{.C}
#include <math.h>
_Bool const inf =  (1.0/0.0 == INFINITY);
```

This is maybe transitional and known (all my apologies) but it is suspicious that it only appears in with the combination of trunk and C23, were normally this should be valid code, regardless of the C version.

Thanks
Jens

### PR DESCRIPTION
**Pull request by Sirraide**
Consider the following code:
```c
bool const inf =  (1.0/0.0);
``` 

When trying to emit the initialiser of this variable in C23, we end up hitting a code path in codegen in `VarDecl::evaluateValueImpl()` where we check for `IsConstantInitialization && (Ctx.getLangOpts().CPlusPlus || Ctx.getLangOpts().C23)`, and if that is the case and we emitted any notes, constant evaluation fails, and as a result, codegen issues this error:
```
<source>:1:12: error: cannot compile this static initializer yet
    1 | bool const inf =  (1.0/0.0);
      |        
```

As a fix, only fail in C23 mode if we’re initialising a `constexpr` variable.

This fixes #84784.

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -259,6 +259,10 @@ Bug Fixes in This Version
   operator.
   Fixes (#GH83267).
 
+- Clang no longer fails to codegen static ``const`` variables whose initialiser performs
+  a floating-point division by 0 in C23.
+  Fixes (#GH84784).
```
Comment (efriedma-quic) [inline clang/docs/ReleaseNotes.rst:264]: We usually only add release notes for things that are fixed between releases.  This wasn't broken in 18, so I don't think we need a release note.

Comment (Sirraide) [reply clang/docs/ReleaseNotes.rst:264]: Oh, I thought it was, my bad. Yeah, we don’t need a release note then ofc.

**Review by AaronBallman — COMMENTED** [review]
Thank you for the fix! This is looking pretty good to me.

File: clang/test/CodeGen/const-init.c [hunk 2]
```diff
@@ -216,3 +216,6 @@ int PR4517_x2 = PR4517_arrc[PR4517_idx];
 // CHECK: @PR4517_x = global i32 42, align 4
 // CHECK: @PR4517_idx = constant i32 1, align 4
 // CHECK: @PR4517_x2 = global i32 42, align 4
+
+// CHECK: @GH84784_inf = constant i8 1
+_Bool const GH84784_inf = (1.0/0.0);
```
[the remark this answers is not in the document]
Comment (Sirraide) [reply clang/test/CodeGen/const-init.c:221]: Iirc we already have `constexpr` tests somewhere, but probably none for that exact code, so we should probably do that yeah.

**Comment by Sirraide** [comment]
I’ve added the test. The CI failure is the same thing from a few days ago I think; must be because I created this branch before updating the main branch.

File: clang/test/Sema/const-init.c [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c23 %s
+
+// Division by 0 here is an error iff the variable is 'constexpr'.
+const _Bool inf1 =  (1.0/0.0 == __builtin_inf());
+constexpr _Bool inf2 = (1.0/0.0 == __builtin_inf()); // expected-error {{must be initialized by a constant expression}} expected-note {{division by zero}}
+constexpr _Bool inf3 = __builtin_inf() == __builtin_inf();
```
Comment (AaronBallman) [inline clang/test/Sema/const-init.c:6]: Please add a newline to the end of the file.

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/test/Sema/const-init.c (hunk 3 above)
Comment (Sirraide) [reply clang/test/Sema/const-init.c]: And here I was thinking that I’d finally stopped forgetting about this haha...

**Comment by Sirraide** [comment]
CI is just going to fail again because of the aforementioned problem, so I’m just going to merge this now and hope I won’t have to revert it later.

**Pull request merged**